# Import Libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os


In [2]:
path=r'C:\Users\k_pat\OneDrive\Documents\Instacart Basket Analysis'

In [3]:
customers = pd.read_csv(os.path.join(path, '2. Data', 'Original Data', 'customers.csv'))

In [4]:
customers

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [5]:
customers.shape

(206209, 10)

# 2. Data Wrangling

In [6]:
#renaming columns
customers.rename(columns= {'First Name' : 'first_name', 'Surnam' : 'surname','Gender' : 'gender','STATE' : 'state',
                          'Age' : 'age', 'n_dependants': 'no_of_dependants', 'fam_status' : 'marital_status'}, inplace= True)

In [7]:
customers.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   no_of_dependants marital_status  income  
0                 3        married  165665  
1                 0         single   59285  
2                 2        married   99568  
3                 0         single   42049  
4                 1        married   40374

# 3. Data Consistency Check

In [8]:
customers.describe()

user_id            age  no_of_dependants         income
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

In [9]:
# checking data types
customers.dtypes

user_id              int64
first_name          object
surname             object
gender              object
state               object
age                  int64
date_joined         object
no_of_dependants     int64
marital_status      object
income               int64
dtype: object

In [10]:
#checking for the missing values
customers.isnull().sum()

user_id                 0
first_name          11259
surname                 0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependants        0
marital_status          0
income                  0
dtype: int64

In [11]:
customers[customers['first_name'].isnull()==True]

user_id first_name   surname  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        no_of_dependants marital_status  income  
53                     2        married   41709  
73                     0         single   82518  
82                     3        married  117099  
99                     1        married  155673  
105                    3        married  151819  
...                  ...            ...     ...  
206038                 3        married   87783  
206044                 2        married  108488  
206090                 1        married   45275  
206162                 0         single   56325  
206171                 2        married   59222  

[11259 rows x 10 columns]

11259 missing values missing for first_name. I don't think this will have much impact on the analysis as we have user_id to identify customers. I will leave the missing values.

In [12]:
# Checking for mixed data types 
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

first_name


# first_name has mixed data type

In [13]:
# Convert first_name data type to string
customers['first_name']=customers['first_name'].astype ('str')

In [14]:
# re-run data types check
customers.dtypes

user_id              int64
first_name          object
surname             object
gender              object
state               object
age                  int64
date_joined         object
no_of_dependants     int64
marital_status      object
income               int64
dtype: object

In [15]:
# Checking for mixed data types 
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

# No other mixed data types

In [16]:
customers.shape

(206209, 10)

In [17]:
# Checking for duplicates
customers[customers.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, no_of_dependants, marital_status, income]
Index: []

# 4. Reducing data types to save RAM before merge

In [18]:
customers['age'] = customers['age'].astype('int8')
customers['no_of_dependants'] = customers['no_of_dependants'].astype('int8')
customers['income'] = customers['income'].astype('int32')
customers['user_id'] = customers['user_id'].astype('str')

In [19]:
customers.dtypes

user_id             object
first_name          object
surname             object
gender              object
state               object
age                   int8
date_joined         object
no_of_dependants      int8
marital_status      object
income               int32
dtype: object

# 5. Merge data with ords_prods data

In [20]:
# Import orders_products_merged pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, '2. Data', 'Prepared Data', 'orders_products_merged_updated1.pkl'))

In [21]:
ords_prods_merge.shape

(32404859, 24)

In [22]:
ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype('str')

In [23]:
ords_prods_merge.dtypes

order_id                            int64
user_id                            object
order_number                        int64
orders_day_of_week                  int64
order_hour_of_day                   int64
days_since_prior_order            float64
product_id                          int64
add_to_cart_order                   int64
reordered                           int64
product_name                       object
aisle_id                            int64
department_id                       int64
prices                            float64
_merge                           category
price_range_loc                    object
busiest_day                        object
busiest_days                       object
busiest_period_of_day              object
max_order                           int64
loyalty_flag                       object
avg_price                         float64
spender_flag                       object
median_days_since_prior_order     float64
frequency_flag                    

In [24]:
ords_prods_custs = ords_prods_merge.merge(customers, on = 'user_id')

In [25]:
ords_prods_custs.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...         frequency_flag  first_name  surname  gender  \
0         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   

     state age date_joined no_of_dependants  marital_status income  
0  Alabama  31   2/17/2019                3         married  40423  
1  Alabama  31   2/17/2019                3         married  40423  
2  Alabama  31   2/17/2019                3         married  40423  
3  Alabama  31   2/17/2019                3         married  40423  
4  Alabama  31   2/17/2019                3         married  40423  

[5 rows x 33 columns]

In [26]:
# Checking shape.  9 new columns is correct, number of rows is still the same.
ords_prods_custs.shape

(32404859, 33)

In [27]:
ords_prods_custs = ords_prods_merge.merge(customers, on = 'user_id', how = 'outer', indicator = True)

ValueError: Cannot use name of an existing column for indicator column

In [28]:
# Export new merged file as pkl
ords_prods_custs.to_pickle(os.path.join(path, '2. Data','Prepared Data', 'full_instacart_data.pkl'))